sale_order_merge/models/sale_order.py

```diff
@@ -0,0 +1,74 @@
+# coding: utf-8
+# Copyright 2016 Opener B.V. - Stefan Rijnhart
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from openerp import api, fields, models
+from openerp.tools.translate import _
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    # Out of scope: exception states ('invoice_except', 'shipping_except')
+    MERGE_STATES = ['draft', 'sent', 'confirm', 'waiting_date',
+                    'progress', 'manual']
+
+    @api.multi
+    @api.depends('merge_with')
+    def _compute_merge_ok(self):
+        for sale in self:
+            sale.merge_ok = bool(sale.merge_with)
+
+    @api.multi
+    def _compute_merge_with(self):
+        for sale in self:
+            sale.merge_with = self.search([('merge_with', '=', sale.id)])
+
+    def _search_merge_with(self, op, arg):
+        """ Apply criteria with which other sale orders the given order
+        is mergeable. """
+        if op != '=' and not arg or not isinstance(arg, (int, long)):
+            return [('id', '=', False)]
+        sale = self.browse(arg)
+        if sale.state not in self.MERGE_STATES or not sale.order_line:
+            return [('id', '=', False)]
+        policy_clause = []
+        if sale.state not in ('draft', 'sent'):
+            policy_clause = [
+                '|', ('order_policy', '=', sale.order_policy),
+                ('state', 'in', ('draft', 'sent'))]
+
+        domain = [
+            ('id', '!=', arg),
+            ('partner_id', '=', sale.partner_id.id),
+            ('partner_shipping_id', '=', sale.partner_shipping_id.id),
+            ('warehouse_id', '=', sale.warehouse_id.id),
+            ('state', 'in', self.MERGE_STATES),
+        ] + policy_clause
```

Comment:
Looking towards extensibility, please build the domain in a separate method. For example, in sale_operating_unit (https://github.com/OCA/operating-unit/pull/22) you would not want to merge sales orders that belong to separate operating units. Also, why is the company_id not included?